### PR TITLE
PulseAudio implementaion for full-duplex

### PR DESCRIPTION
**Pull request by achronop**
This implement the full-duplex functionality for PulseAudio back-end in the way that described in kinetiknz/cubeb#63

Test files has been updated as required and the test-suite runs successfully under Linux/Pulse.

### CODE REVIEW

File: include/cubeb/cubeb.h [hunk 1]
```diff
@@ -172,8 +175,6 @@ typedef enum {
 
 /** An opaque handle used to refer a particular input or output device
  *  accross calls. */
```
Comment (kinetiknz) [inline include/cubeb/cubeb.h]: Move the comment with the code.

File: src/cubeb_alsa.c [hunk 2]
```diff
@@ -85,7 +85,9 @@ struct cubeb_stream {
   snd_pcm_uframes_t write_position;
   snd_pcm_uframes_t last_position;
   snd_pcm_uframes_t buffer_size;
-  cubeb_stream_params params;
+  snd_pcm_uframes_t period_size;
```
Comment (kinetiknz) [inline src/cubeb_alsa.c]: I can't see a reason for this being reintroduced.  Looks like a merge error has reverted a change made on master.  This applies to every other occurrence of period_size in the changeset.

Comment (kinetiknz) [inline src/cubeb_alsa.c]: I'd rather see the ALSA changes separately.  I don't think they should be part of this PR.

File: src/cubeb_pulse.c [hunk 3]
```diff
@@ -170,46 +188,52 @@ stream_state_callback(pa_stream * s, void * u)
...
 
   towrite = nbytes;
-
+  read_offset = 0;
   while (towrite) {
     size = towrite;
     r = WRAP(pa_stream_begin_write)(s, &buffer, &size);
+    if (r < 0) {
+      // Never get here in normal scenario
+      LOG("Unexpected error. Debugging with rr causes this\n");
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: This needs to be tracked down.  What error are you seeing returned here?

The docs don't seem to have any documented error codes, but looking at pulseaudio/src/pulse/stream.c, it can return PA_ERR_FORKED, PA_ERR_BADSTATE, or PA_ERR_INVALID.

File: src/cubeb_pulse.c [hunk 4]
```diff
@@ -170,46 +188,52 @@ stream_state_callback(pa_stream * s, void * u)
...
+      WRAP(pa_stream_cancel_write)(s);
+      stm->shutdown = 1;
+      return;
+    }
     assert(r == 0);
     assert(size > 0);
     assert(size % frame_size == 0);
 
-    got = stm->data_callback(stm, stm->user_ptr, NULL, buffer, size / frame_size);
+    LOG("Trigger user callback with output buffer size %zd, read_offset=%zd\n", size, read_offset);
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: Be consistent about "var=%" or "var %" in logging.

File: src/cubeb_pulse.c [hunk 5]
```diff
@@ -247,6 +271,79 @@ stream_request_callback(pa_stream * s, size_t nbytes, void * u)
...
 
 static int
+read_from_input(pa_stream * s, const void ** buffer, size_t * size)
+{
+  if (WRAP(pa_stream_readable_size)(s) > 0) {
+    if (WRAP(pa_stream_peek)(s, buffer, size) < 0) {
+      return -1;
+    }
+  }
+  return WRAP(pa_stream_readable_size)(s);
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: Can you explain the logic here? I don't understand why pa_stream_readable_size is being called twice.

File: src/cubeb_pulse.c [hunk 6]
```diff
@@ -247,6 +271,79 @@ stream_request_callback(pa_stream * s, size_t nbytes, void * u)
...
+{
+  LOG("Input callback buffer size %zd\n", nbytes);
+
+  cubeb_stream * stm;
+  stm = u;
+  if (stm->shutdown) {
+    return;
+  }
+
+  const void *read_data=NULL;
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: Spaces: const void * read_data = NULL;

File: src/cubeb_pulse.c [hunk 7]
```diff
@@ -247,6 +271,79 @@ stream_request_callback(pa_stream * s, size_t nbytes, void * u)
...
+  size_t read_size;
+  while (read_from_input(s, &read_data, &read_size) > 0) {
+    size_t in_frame_size = WRAP(pa_frame_size)(&stm->input_sample_spec);
+    size_t read_frames = read_size / in_frame_size;
+
+    cubeb_stream * stm = u;
+    if (stm->output_stream) {
+      // input/capture + output/playback operation
+      void * read_buffer = (void*)read_data;
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: Spaces: (void *) read_data;

File: src/cubeb_pulse.c [hunk 8]
```diff
@@ -170,46 +188,52 @@ stream_state_callback(pa_stream * s, void * u)
...
+      stm->shutdown = 1;
+      return;
+    }
     assert(r == 0);
     assert(size > 0);
     assert(size % frame_size == 0);
 
-    got = stm->data_callback(stm, stm->user_ptr, NULL, buffer, size / frame_size);
+    LOG("Trigger user callback with output buffer size %zd, read_offset=%zd\n", size, read_offset);
+    got = stm->data_callback(stm, stm->user_ptr, (uint8_t *)input_data + read_offset, buffer, size / frame_size);
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: Spaces: (uint8_t *) input_data

File: src/cubeb_pulse.c [hunk 9]
```diff
@@ -247,6 +271,79 @@ stream_request_callback(pa_stream * s, size_t nbytes, void * u)
...
+    if (stm->output_stream) {
+      // input/capture + output/playback operation
+      void * read_buffer = (void*)read_data;
+      size_t out_frame_size = WRAP(pa_frame_size)(&stm->output_sample_spec);
+      size_t write_size = read_frames * out_frame_size;
+      // Offer full duplex data for writing
+      trigger_user_callback(stm->output_stream, read_buffer, write_size, stm);
+    } else {
+      // input/capture only operation. Call callback directly
+      void * read_buffer = (void*)read_data;
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: Ditto.

File: src/cubeb_pulse.c [hunk 10]
```diff
@@ -247,6 +271,79 @@ stream_request_callback(pa_stream * s, size_t nbytes, void * u)
...
+  stm = u;
+  if (stm->shutdown) {
+    return;
+  }
+
+  const void *read_data=NULL;
+  size_t read_size;
+  while (read_from_input(s, &read_data, &read_size) > 0) {
+    size_t in_frame_size = WRAP(pa_frame_size)(&stm->input_sample_spec);
+    size_t read_frames = read_size / in_frame_size;
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: Since read_frames is passed as a long, it should be declared as a long and any computations checked to ensure they fit into a long.

File: src/cubeb_pulse.c [hunk 11]
```diff
@@ -247,6 +271,79 @@ stream_request_callback(pa_stream * s, size_t nbytes, void * u)
...
+      void * read_buffer = (void*)read_data;
+      size_t out_frame_size = WRAP(pa_frame_size)(&stm->output_sample_spec);
+      size_t write_size = read_frames * out_frame_size;
+      // Offer full duplex data for writing
+      trigger_user_callback(stm->output_stream, read_buffer, write_size, stm);
+    } else {
+      // input/capture only operation. Call callback directly
+      void * read_buffer = (void*)read_data;
+      long got = stm->data_callback(stm, stm->user_ptr, read_buffer, NULL, read_frames);
+      size_t u_got = got;
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: If read_frames was a long, as mentioned above, this wouldn't be necessary.

File: src/cubeb_pulse.c [hunk 12]
```diff
@@ -247,6 +271,79 @@ stream_request_callback(pa_stream * s, size_t nbytes, void * u)
...
+      if ( got < 0 || u_got != read_frames) {
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: No space: (got

File: src/cubeb_pulse.c [hunk 13]
```diff
@@ -290,16 +404,25 @@ operation_wait(cubeb * ctx, pa_stream * stream, pa_operation * o)
 }
 
 static void
-stream_cork(cubeb_stream * stm, enum cork_state state)
+cork_io_stream(cubeb_stream * stm, pa_stream * io_stream, enum cork_state state)
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: I'm not sure I see the value in splitting stream_cork (and wait_until_stream_ready) up like this.

File: src/cubeb_pulse.c [hunk 14]
```diff
@@ -290,16 +404,25 @@ operation_wait(cubeb * ctx, pa_stream * stream, pa_operation * o)
...
 {
-  pa_operation * o;
-
-  WRAP(pa_threaded_mainloop_lock)(stm->context->mainloop);
-  o = WRAP(pa_stream_cork)(stm->stream, state & CORK, stream_success_callback, stm);
+  pa_operation * o = NULL;
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: No need to initialize this, it's clearly initialized on first use.

File: src/cubeb_pulse.c [hunk 15]
```diff
@@ -485,91 +636,153 @@ pulse_destroy(cubeb * ctx)
...
        until some point after initialization has completed. */
-    o = WRAP(pa_stream_update_timing_info)(stm->stream, stream_success_callback, stm);
-    if (o) {
-      r = operation_wait(stm->context, stm->stream, o);
-      WRAP(pa_operation_unref)(o);
-    }
+    r = stream_update_timing_info(stm);
   }
+
+#ifdef LOGGING_ENABLED
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: It seems like it'd be better to log this after dropping the lock and testing r != 0.

File: src/cubeb_resampler.h [hunk 16]
```diff
@@ -54,7 +54,7 @@ cubeb_resampler * cubeb_resampler_create(cubeb_stream * stream,
  */
 long cubeb_resampler_fill(cubeb_resampler * resampler,
                           void * input_buffer, 
-						  void * output_buffer, long frames_needed);
+                          void * output_buffer, long frames_needed);
```
Comment (kinetiknz) [inline src/cubeb_resampler.h]: This change and the test changes should be in a separate PR as well.

File: src/cubeb_pulse.c [hunk 17]
```diff
@@ -308,6 +431,34 @@ stream_cork(cubeb_stream * stm, enum cork_state state)
...
+{
+  int r = -1;
+  pa_operation * o = NULL;
+  if (stm->output_stream) {
+    o = WRAP(pa_stream_update_timing_info)(stm->output_stream, stream_success_callback, stm);
+    if (o) {
+      r = operation_wait(stm->context, stm->output_stream, o);
+      WRAP(pa_operation_unref)(o);
+    }
+    // if this fail abort
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: Comment seems unnecessary.

File: src/cubeb_pulse.c [hunk 18]
```diff
@@ -485,91 +636,153 @@ pulse_destroy(cubeb * ctx)
 
 static void pulse_stream_destroy(cubeb_stream * stm);
 
+pa_sample_format_t
+cube_to_pulse_format(cubeb_sample_format format)
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: cubeb not cube

File: src/cubeb_pulse.c [hunk 19]
```diff
@@ -485,91 +636,153 @@ pulse_destroy(cubeb * ctx)
...
+    return PA_SAMPLE_FLOAT32BE;
+  default:
+    return PA_SAMPLE_INVALID;
+  }
+}
+
+static pa_stream *
+create_pa_stream(cubeb_stream * stm, cubeb_stream_params * stream_params, char const * stream_name)
+{
+  assert(stm || stream_params);
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: Should be &&

File: src/cubeb_pulse.c [hunk 20]
```diff
@@ -485,91 +636,153 @@ pulse_destroy(cubeb * ctx)
...
+  pa_sample_spec ss;
+  ss.format = cube_to_pulse_format(stream_params->format);
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: If cubeb_to_pulse_format returns PA_SAMPLE_INVALID, we still try to create a stream.  We also lose the old CUBEB_ERROR_INVALID_FORMAT result from pulse_stream_init which could be used to try fallback formats.

File: src/cubeb_pulse.c [hunk 21]
```diff
@@ -485,91 +636,153 @@ pulse_destroy(cubeb * ctx)
...
+                                     PA_STREAM_START_CORKED | PA_STREAM_ADJUST_LATENCY,
+                                     NULL, NULL);
   }
-  WRAP(pa_stream_set_state_callback)(stm->stream, stream_state_callback, stm);
-  WRAP(pa_stream_set_write_callback)(stm->stream, stream_request_callback, stm);
-  WRAP(pa_stream_connect_playback)(stm->stream, NULL, &battr,
+
+  // Set up input stream
+  if (input_stream_params) {
+    stm->input_stream = create_pa_stream(stm, input_stream_params, "cubeb input");
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: This should use the same stream name as the output so they're paired correctly.  If we can't use an identical string, appending "(input)" or something to stream_name would be better than using a generic name.

File: src/cubeb_pulse.c [hunk 22]
```diff
@@ -485,91 +636,153 @@ pulse_destroy(cubeb * ctx)
...
+    WRAP(pa_stream_set_read_callback)(stm->input_stream, stream_read_callback, stm);
+
+    battr = set_buffering_attribute(latency, &stm->input_sample_spec);
+    WRAP(pa_stream_connect_record)(stm->input_stream,
+                                   input_stream_params->devid,
+                                   &battr,
                                    PA_STREAM_AUTO_TIMING_UPDATE | PA_STREAM_INTERPOLATE_TIMING |
-                                   PA_STREAM_START_CORKED,
-                                   NULL, NULL);
+                                   PA_STREAM_NOT_MONOTONIC |
```
Comment (kinetiknz) [inline src/cubeb_pulse.c]: NOT_MONOTONIC is only documented to affect the result of pa_stream_get_time, which I can't see ever being called on the input_stream.  Why is the flag being used here?

File: include/cubeb/cubeb.h [hunk 23]
```diff
@@ -127,6 +129,7 @@ typedef struct {
 #if defined(__ANDROID__)
   cubeb_stream_type stream_type; /**< Used to map Android audio stream types */
 #endif
+  cubeb_devid devid;
```
Comment (kinetiknz) [inline include/cubeb/cubeb.h]: At the very least: this needs to be documented.

I don't like adding this to stream_params, because it's pretty dangerous if the caller forgets to initialize it (which we've already seen happen during integration with Gecko), and we have no way to sanity check this value unlike the other values in this struct.

I think the original plan was to pass a devid in as a new parameter to cubeb_stream_params.  I guess one each for the input and output stream.  At least it's impossible to forget to pass a value in that case, and we're already breaking the API by adding the input stream parameters.

**Comment by kinetiknz** [comment]
Thanks for working on this Alex.

I think for merging the final changes I'd prefer to see the history squashed so each commit contains a single logic change, etc. rather than pulling in the development history.

File: src/cubeb_pulse.c (hunk 3 above)
Comment (achronop) [inline src/cubeb_pulse.c]: I am afraid the story is more complicated here. The bug experienced once by jesup while debugging with rr. The problem was that write stream was in state PA_STREAM_CREATING (so I guess the error was PA_ERR_BADSTATE) which should not be possible because we wait until stream is in PA_STREAM_READY state inside stream_init. Unfortunately I am not able to reproduce it here in order to track it and I prefer to check and report than failing silently.

File: src/cubeb_pulse.c (hunk 5 above)
Comment (achronop) [inline src/cubeb_pulse.c]: I thought it's cheap enough to call it again. I see it is confusing though and I will change it to call it just once.

File: src/cubeb_pulse.c (hunk 9 above)
Comment (achronop) [inline src/cubeb_pulse.c]: This is another API change for cubeb. Since we have also to deal with devid issue we can add there this change.

File: src/cubeb_pulse.c (hunk 10 above)
Comment (achronop) [inline src/cubeb_pulse.c]: `read_frames` take part to additional computation inside if statement. Instead of changing that i created a method to calculate the size that does not produce overflow. I found it easier to consume all the data in this way. Please take a look and if you think that's not the right approach let me know.

File: src/cubeb_pulse.c (hunk 20 above)
Comment (achronop) [inline src/cubeb_pulse.c]: I updated the code to bring back this functionality.

File: src/cubeb_pulse.c (hunk 21 above)
Comment (achronop) [inline src/cubeb_pulse.c]: I use the same name now and seems to work. The documentation also does not mention anything about unique name.

**Comment by achronop** [comment]
Thank you very much for your time!

I have updated and/or comment back most of the comments. I've left behind those that will change due to interface changes. They will be updated by rebase. Unfortunately check fails now.

I've also left squash for then. When we have API changes landed I will fix this too. I know it looks ugly now, sorry for that.

In the mean time feel free to comment on anything of the above. Thanks again!

**Comment by kinetiknz** [comment]
I addressed the my remaining comments, merged and pushed the squashed version here: https://github.com/kinetiknz/cubeb/commit/79ff75467fd439b435843d98da0d124babc6233b
